StylePropertyMap set method implemented

njlawton opened the pull request:
The set method has been implemented for longhand CSS properties that are defined by only one StyleValue

Commit:
Implemented the set method for the StylePropertyMap class
This currently has only been implemented for the most specific longhand values i.e. ones that only support setting one value only
LongHands that are set using multiple values and those that accept sequences of style values are currently unsupported

Commit:
Merge branch 'property-dictionary' into Style-map-set_updated

Commit:
Shortened lines in StylePropertyMap class

Commit:
Merge branch 'master' into Style-map-set_updated

Commit:
Added new test case to the Inline StylePropertyMap test class
This test will see if rubbish inputs (Non keyword value) will throw errors

wilddamon commented on src/style-property-map.js:
```diff
@@ -20,7 +20,26 @@
-    throw new TypeError('Function not implemented yet');
+    var cssPropertyDictionary = propertyDictionary();
+    if (!cssPropertyDictionary.isSupportedProperty(property)) {
+      throw new TypeError(property + 'Is not a Supported CSS property');
```
Nitpick: Need a space and should be a lowercase "i" for "is".

``` javascript
throw new TypeError(property + ' is not a supported CSS property.');
```

wilddamon commented on src/style-property-map.js:
```diff
@@ -20,7 +20,26 @@
+    if (value instanceof StyleValue) {
+      if (!cssPropertyDictionary.isValidInput(property, value)) {
+        throw new TypeError(
+          'This StyleValue type cannot be set to this property');
```
It's better to include the property and value in the error message.

``` javascript
throw new TypeError(property + ' does not take values of type ' + value.constructor.name);
```

wilddamon commented on src/style-property-map.js:
```diff
@@ -20,7 +20,26 @@
+        'Setting a sequence of StyleValues is not implemented yet');
+    }
+    throw new TypeError(
+      'The value you are setting must be a StyleValue object');
```
"... must be a StyleValue or sequence of StyleValues"?

wilddamon commented on test/js/inline-style-property-map.js:
```diff
@@ -8,8 +8,44 @@ suite('Inline StylePropertyMap', function() {
   });
 
-  test('Element.styleMap method returns a StylePropertyMap object', function() {
+  test('the Element.styleMap method returns a StylePropertyMap object for that element', function() {
```
Nitpick: Should start with a capital letter.

wilddamon commented on src/style-property-map.js:
```diff
@@ -20,7 +20,26 @@
+      return;
+    }
+
+    if (value instanceof Array) {
```
How about we put this first, then it can eventually be passed off to a helper function. Then you can get rid of the floating `return` statement and write something like

``` javascript
StylePropertyMap.prototype.set = function(property, value) {
  var cssPropertyDictionary = propertyDictionary();
  if (!cssPropertyDictionary.isSupportedProperty(property)) {
    throw new TypeError(property + ' is not a supported CSS property.');
  }

  if (value instanceof Array) {
    // pass to helper function
  }

  if (!value instanceof StyleValue) {
    throw new TypeError('value must be a StyleValue or sequence of StyleValues');
  }
  if (!cssPropertyDictionary.isValidInput(property, value)) {
    throw new TypeError(property + ' does not take values of type ' + value.constructor.name);
  }

  this._styleObject[property] = value.cssString;
}
```

wilddamon commented on test/js/inline-style-property-map.js:
```diff
@@ -8,8 +8,44 @@ suite('Inline StylePropertyMap', function() {
+    assert.instanceOf(inlineStyleMap, StylePropertyMap);
+  });
+
+  test('The set method should correctly set a supported StyleValue to a CSS style property of an HTML element', function() {
```
This title is a bit confusing, how about

``` javascript
test('The set method successfully sets the CSS string of the StyleValue on an element.', function() {
```

wilddamon commented on test/js/inline-style-property-map.js:
```diff
@@ -8,8 +8,44 @@ suite('Inline StylePropertyMap', function() {
+    var keyword = new KeywordValue('auto');
+    inlineStyleMap.set('height', keyword);
+
+    assert.strictEqual(this.element.style['height'], keyword.cssString);
```
This test is actually testing the same thing as the above test. You can delete it.

wilddamon commented on test/js/inline-style-property-map.js:
```diff
@@ -8,8 +8,44 @@ suite('Inline StylePropertyMap', function() {
+    assert.throw(function() {inlineStyleMap.set('height', numberValue)}, TypeError);
+  });
+
+  test('The set method should throw a TypeError if a KeywordValue unsupported by the CSS style property is set ', function() {
```
A KeywordValue is a StyleValue, so this test is testing the same thing as the one above it. It can also be deleted.

wilddamon commented on test/js/inline-style-property-map.js:
```diff
@@ -8,8 +8,44 @@ suite('Inline StylePropertyMap', function() {
+
+    assert.throw(function() {inlineStyleMap.set('height', 4)}, TypeError);
   });
 });
```
Please also add a test for setting an unsupported property, something like

``` javascript
inlineStyleMap.set('lemons', new SimpleLength(3, 'px'));
```

Commit:
Fixed errors and formatting issues in the StylePropertyMap class and its test class
This included changing some test and error messages to be less confusing as well as changing the order of some logic in the set method to avoid clutter and increase code clarity

njlawton commented:
I have changed how error messages are displayed like you said but due to getting confusing messages like height does not take values of type KeywordValue I have separated out the errors into two so that if an incorrect keyword is entered it will say something like property + ' does not take the keyword ' + value.cssString as a result I have not removed (and slightly updated) the test case you asked me to remove

Commit:
Added test case for trying to put an unsupported property into the set method

Commit:
Fixed small formating errors found with the linter

wilddamon commented:
Oh I see. Perhaps a note in the test name that it is for StyleValues that are not KeywordValues then would clear up any confusion.

Commit:
Changed wording of inline StylePropertyMap test case to avoid confusion

njlawton commented:
As I changed the wording of the test case above it to fix the problem the box is not going away

wilddamon commented:
LGTM

The comment still shows up because you didn't change that line. Which is as expected so no worries :)
